### PR TITLE
#161 Created a setting for allowed duration above max soc

### PR DESCRIPTION
**Pull request by ArdJonker**
Set back_to_soc_max end time based on changing soc / disconnetion Change soc_maxima based on scenario's:
- Calendar item before back_to_soc_max
- No calendar item

### CODE REVIEW

**Comment by ArdJonker** [comment]
It is a big one. Let's review together tomorrow.

File: flexmeasures_client.py [hunk 1]
```diff
@@ -355,25 +343,63 @@ def trigger_schedule(self, *args, **fnc_kwargs):
...
+                # the current SoC is used.
+                # The relaxation window is still based on the CAR_MAX_SOC_IN_KWH.
+                soc_maxima = [
+                    {
+                        "value": current_soc_kwh,
+                        "datetime": dt.isoformat(),
+                    } for dt in [rounded_now + x * resolution for x in range(0, (start_relaxation_window - rounded_now) // resolution) ]
+                ]
+            else:
+                # No calendar item so we keep the current_soc as the maxima until back_to_max_soc.
```
Comment (Flix6x) [inline flexmeasures_client.py:365]: No calendar item **within the discharging window**.

File: flexmeasures_client.py [hunk 2]
```diff
@@ -355,25 +343,63 @@ def trigger_schedule(self, *args, **fnc_kwargs):
...
+                # The only thing that changes is the setting for the soc maxima, instead of the usual CAR_MAX_SOC_IN_KWH
+                # the current SoC is used.
+                # The relaxation window is still based on the CAR_MAX_SOC_IN_KWH.
+                soc_maxima = [
+                    {
+                        "value": current_soc_kwh,
+                        "datetime": dt.isoformat(),
+                    } for dt in [rounded_now + x * resolution for x in range(0, (start_relaxation_window - rounded_now) // resolution) ]
+                ]
+            else:
```
Comment (Flix6x) [inline flexmeasures_client.py:364]: Let's have explicit scenarios for these cases:
- Start of calendar item's charging relaxation window is right of back_to_soc_max: `A < B < C`
- Start of calendar item's charging relaxation window is right of now, but left of back_to_soc_max: : `A < C < B`
- Start of calendar item's charging relaxation window is left of now: `C < A < B`

File: constants.py [hunk 3]
```diff
@@ -1,14 +1,27 @@
...
+CAR_MIN_SOC_IN_KWH: float
+
+# A 'soft' setting, that is respected during normal cycling but is ignored when
+# a calendar item requires a higher SoC.
 # Defaults to 80% (to be safe)
 CAR_MAX_SOC_IN_PERCENT: int = 80
+# Derived from above setting and CAR_MAX_CAPACITY_IN_KWH
+CAR_MAX_SOC_IN_KWH: float
+
+# Duration in hours, defaults to 12 should be between 2 and 36 hours
```
Comment (Flix6x) [inline constants.py:23]: Let's try to support 0 hours and 1 hour, too.

**Review by Flix6x — CHANGES_REQUESTED** [review]

File: README.md [hunk 4]
```diff
@@ -311,10 +311,18 @@ car_min_soc_in_percent: 20
 # The setting must be an integer value between 60 and 100, default is 80.
 car_max_soc_in_percent: 80
 
+# When the car connects with a SoC higher than car_max_soc_in_percent
+# how long may the schedule take to bring the SoC back to this maximum?
+# Higher values give opportunity for higher gains but might have a (minor)
```
[the remark this answers is not in the document]
Comment (ArdJonker) [reply README.md:316]: OK, will change

File: flexmeasures_client.py (hunk 1 above)
Comment (ArdJonker) [reply flexmeasures_client.py:365]: Will clarify

File: flexmeasures_client.py [hunk 5]
```diff
@@ -335,45 +323,131 @@ def trigger_schedule(self, *args, **fnc_kwargs):
...
+                        step_kwh = (current_soc_kwh - c.CAR_MAX_SOC_IN_KWH) / number_of_steps
+                        i = 0
+                        while i < number_of_steps:
+                            d = {
+                                "value": current_soc_kwh - (i * step_kwh),
+                                "datetime": (rounded_now + i * resolution).isoformat()
+                            }
+                            self.log(f"dict: {d}")
+                            soc_maxima_higher_max_soc.append(d)
+                            i += 1
```
Comment (Flix6x) [inline flexmeasures_client.py:418]: `soc_maxima_higher_max_soc += [{<dict you assigned to d>} for i in range(number_of_steps)]`

File: v2g_liberty.py [hunk 6]
```diff
@@ -604,6 +614,15 @@ def set_next_action(self):
...
             return
 
+        # If the SoC of the car is higher than the max-soc (intended for battery protection)
+        # a target is to return to the max-soc within the ALLOWED_DURATION_ABOVE_MAX_SOC
+        if (self.back_to_max_soc is None) and (self.connected_car_soc_kwh > c.CAR_MAX_SOC_IN_KWH):
+            self.back_to_max_soc = time_round((self.get_now() + timedelta(hours=c.ALLOWED_DURATION_ABOVE_MAX_SOC)), self.MIN_RESOLUTION)
+            self.log(f"SoC above max-soc, aiming to schedule with target {c.CAR_MAX_SOC_IN_PERCENT}% at {self.back_to_max_soc}.")
+        elif self.connected_car_soc_kwh <= c.CAR_MAX_SOC_IN_KWH:
+            self.back_to_max_soc = None
+            self.log(f"SoC was below max-soc, has been restored.")
```
[the remark this answers is not in the document]
Comment (ArdJonker) [reply v2g_liberty.py:624]: The idea is that both conditions only occur once in a "cycle" of the SoC being > max. So, indeed the extra condition is needed for that.

File: v2g_liberty.py [hunk 7]
```diff
@@ -33,6 +34,10 @@ class V2Gliberty(hass.Hass, WallboxModbusMixin):
     connected_car_soc: int
     connected_car_soc_kwh: float
 
+    # This is a target datetime at which te SoC that is above the max_soc must return back to or below this value.
+    # It is dependant on the user setting for allowed duration above max soc.
```
[the remark this answers is not in the document]
Comment (ArdJonker) [reply v2g_liberty.py:38]: 🫢

File: README.md [hunk 8]
```diff
@@ -311,10 +311,18 @@ car_min_soc_in_percent: 20
...
+# degradation effect on the battery.
+# The setting must be an integer value between 2 and 36, default is 12.
+allowed_duration_above_max_soc_in_hrs: 12
```
[the remark this answers is not in the document]
Comment (ArdJonker) [reply README.md:319]: Done

File: flexmeasures_client.py [hunk 9]
```diff
@@ -335,45 +323,131 @@ def trigger_schedule(self, *args, **fnc_kwargs):
...
+        #
+        # Assume:
+        # CTM  = Charge Target Moment which is the start of the first upcoming calendar item.
+        #        By default if there is no calendar item, the CTM is one week from now. This gives the
+        #        schedule enough freedom for the coming 27 hours (total duration of the schedule).
+        # SRW  = Start of the relaxation window for the CTM, including the slack of 1 hour.
+        #        Only relevant for calendar items with a target SoC above the CAR_MAX_SOC_IN_KWH.
+        #        Relaxation refers to the fact that in this window the schedule does not get soc-maxima so that
+        #        it can charge above the CAR_MAX_SOC_IN_KWH to reach the higher target SoC.
+        #        To keep things simple SRW remains based on CAR_MAX_SOC_IN_KWH if there is B2MS, not the current higher soc.
```
[the remark this answers is not in the document]
Comment (ArdJonker) [reply flexmeasures_client.py:350]: OK, will adapt.

File: flexmeasures_client.py [hunk 10]
```diff
@@ -335,45 +323,131 @@ def trigger_schedule(self, *args, **fnc_kwargs):
...
+        # B2MS = The datetime at which the ALLOWED_DURATION_ABOVE_MAX_SOC ends, it cannot be in the past.
+        #        It serves as a target with a maximum SoC (where regular targets have a minimum).
+        #        The CTM has a higher priority than the B2MS.
+        # EMDW = End of Minimum Discharge Window. Minimum Discharge Window (MDW) = time needed to discharge from current
```
[the remark this answers is not in the document]
Comment (ArdJonker) [reply flexmeasures_client.py:354]: No, it usually is much nearer in time. EMDW relates to only the minimum time for discharging from a Higher-SoC to Max-SoC. It is usually 1 to 2 hours or so.
B2MS is usually much further in time as the ALLOWED_DURATION_ABOVE_MAX_SOC is by default 12 hours. But they can become swapped around in time as the B2MS remains static relative to the moment where the SoC > max-soc occurred and the EMDW is always relative to "now".

File: flexmeasures_client.py [hunk 11]
```diff
@@ -335,45 +323,131 @@ def trigger_schedule(self, *args, **fnc_kwargs):
...
+        #
+        # The following scenarios need to be handled, they might in time flow from one into the other:
+        # 0. No B2MS
+        #    The soc-maxima are based on the CAR_MAX_SOC_IN_KWH and run from "now" up to SRW.
+        # 1. NOW < B2MS < SRW < CTM
+        #    The B2MS is not influenced by the first calendar item (or there is none)
+        #    SoC maxima are gradually lowered from current soc until B2MS from where they are set to CAR_MAX_SOC_IN_KWH.
+        #    TODO: A drawback of the gradual approach is that there might be discharging with low power which usually is
+        #          less efficient. So, if the trigger message could handle the concept "only discharge during this window"
+        #          it would result in better schedules. This should then replace the gradually lowered soc_maxima.
```
[the remark this answers is not in the document]
Comment (ArdJonker) [reply flexmeasures_client.py:366]: I think that is the best way to go. Let's discuss f2f next week.

File: flexmeasures_client.py (hunk 5 above)
Comment (ArdJonker) [reply flexmeasures_client.py:418]: Better! More elegant!

**Review by Flix6x — APPROVED** [review]

**Pull request merged**